File: usb_android/android/src/main/kotlin/dev/yanshouwang/usb/MyUsbDevice.kt
package dev.yanshouwang.usb

import android.hardware.usb.UsbDevice

class MyUsbDevice(private val collector: MyCollector) : MyUsbDeviceHostApi {
    override fun getDeviceClass(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.deviceClass.toLong()
    }

    override fun getDeviceSubClass(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.deviceSubclass.toLong()
    }

    override fun getDeviceProtocol(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.deviceProtocol.toLong()
    }

    override fun getDeviceName(hashCode: Long): String {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.deviceName
    }

    override fun getVendorId(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.vendorId.toLong()
    }

    override fun getProductId(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.productId.toLong()
    }

    override fun getManufacturerName(hashCode: Long): String? {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.manufacturerName
    }

    override fun getProductName(hashCode: Long): String? {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.productName
    }

    override fun getConfigurationCount(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.configurationCount.toLong()
    }

    override fun getInterfaceCount(hashCode: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.interfaceCount.toLong()
    }

    override fun getVersion(hashCode: Long): String {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.version
    }

    override fun getSerialNumber(hashCode: Long): String? {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        return usbDevice.serialNumber
    }

    override fun getConfiguration(hashCode: Long, index: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        val usbConfiguration = usbDevice.getConfiguration(index.toInt())
        collector.allocate(usbConfiguration)
        return usbConfiguration.hashCode().toLong()
    }

    override fun getInterface(hashCode: Long, index: Long): Long {
        val usbDevice = collector.instanceOf(hashCode) as UsbDevice
        val usbInterface = usbDevice.getInterface(index.toInt())
        collector.allocate(usbInterface)
        return usbInterface.hashCode().toLong()
    }
}

File: usb_android/lib/src/my_usb_device.dart
import 'package:usb_platform_interface/usb_platform_interface.dart';

import 'my_api.dart';
import 'my_object.dart';
import 'my_usb_configuration.dart';
import 'my_usb_interface.dart';

final myUsbDeviceApi = MyUsbDeviceHostApi();

class MyUsbDevice extends MyObject implements UsbDevice {
  MyUsbDevice(super.hashCode);

  @override
  Future<UsbConfiguration> getConfiguration(int index) async {
    final usbConfigurationHashCode =
        await myUsbDeviceApi.getConfiguration(hashCode, index);
    return MyUsbConfiguration(usbConfigurationHashCode);
  }

  @override
  Future<int> getConfigurationCount() {
    return myUsbDeviceApi.getConfigurationCount(hashCode);
  }

  @override
  Future<int> getDeviceClass() {
    return myUsbDeviceApi.getDeviceClass(hashCode);
  }

  @override
  Future<String> getDeviceName() {
    return myUsbDeviceApi.getDeviceName(hashCode);
  }

  @override
  Future<int> getDeviceProtocol() {
    return myUsbDeviceApi.getDeviceProtocol(hashCode);
  }

  @override
  Future<int> getDeviceSubClass() {
    return myUsbDeviceApi.getDeviceSubClass(hashCode);
  }

  @override
  Future<UsbInterface> getInterface(int index) async {
    final usbInterfaceHashCode =
        await myUsbDeviceApi.getInterface(hashCode, index);
    return MyUsbInterface(usbInterfaceHashCode);
  }

  @override
  Future<int> getInterfaceCount() {
    return myUsbDeviceApi.getInterfaceCount(hashCode);
  }

  @override
  Future<String?> getManufacturerName() {
    return myUsbDeviceApi.getManufacturerName(hashCode);
  }

  @override
  Future<int> getProductId() {
    return myUsbDeviceApi.getProductId(hashCode);
  }

  @override
  Future<String?> getProductName() {
    return myUsbDeviceApi.getProductName(hashCode);
  }

  @override
  Future<String?> getSerialNumber() {
    return myUsbDeviceApi.getSerialNumber(hashCode);
  }

  @override
  Future<int> getVendorId() {
    return myUsbDeviceApi.getVendorId(hashCode);
  }

  @override
  Future<String> getVersion() {
    return myUsbDeviceApi.getVersion(hashCode);
  }
}
